### ISSUE TITLE
Add Schema Versioning Discussion

### ISSUE THREAD

**Issue by reinecke** [sandflow/otio-core-specification#3]
The specification should include some description of schema versioning and expected behavior from implementations around upconversion.

**Comment by palemieux** [comment]
@reinecke 

What is the expected behavior from implementations around upconversion?

I would think that automatic upconversion might result in surprises. For example, if implementation B were to automatically upconvert the files of implementation A, then these files may no longer be readable by implementation A.

What about stating that:

* processors should not use `deprecated` entities in new object instances, unless compatibility legacy implementations is specifically required
* processors are expected to accept all deprecated entities

**Comment by reinecke** [comment]
The rules as they exist in the current implementation are:
1. On deserialization, upconvert "old" schema objects to the current version
2. On deserialization, if a known schema name with a newer than supported schema version is encountered, error out

This was to allow users of the API (as distinct from the file format) to be insulated from serialization schema versions and avoid branching code on schema version or potentially having WET (Write Everything Twice, the opposite of DRY) codebases where similar upgrade logic gets strewn across application code.

Given that we are shifting the format toward integrating with external applications where there could be quite a lot of disparity in adoption of the library version (assuming they even choose to use the reference implementation), I see the intent of your proposed rule set. I think there may still be some issues that might cause quite a bit of confusion for users.

One example could be a workflow with _DCC A_ which supports _schema version 1_, and _DCC B_ which supports _schema version 2_. If _DCC B_ loads a file from _DCC A_ and writes it back out with no substantial changes, then _DCC A_ would be able to read that file (no up conversion). If _DCC B_ writes a new file from scratch, then _DCC A_ would not be able to read it (new entities are used). This would likely be unpredictable behavior for users.

More pragmatically, many of the DCCs will deserialize OTIO into their own internal data models and dispose of OTIO-specific data models, likely losing any schema versioning info in the process. Should they write the data back out, they will most likely re-build OTIO data structures for serialization from scratch.

Another thing to consider is that use cases where an OTIO file is read and written back out there is likely to be mutations applied to the data model, some of which may force up conversion of schema. That could leave a file with multiple versions of the same schema. Awkward as it is, I'm not sure how problematic this is in practice.

Without question, the specification should include a description of the migration rules to convert from the immediately previous schema version to the new schema version when the specification is amended to include new versions. Implementations can then execute these rules in order to increment the schema up to the required version. The most common cases of this so far have been either specifying default values for added attributes or remapping of old attribute names to new attribute names. One such example of this is upgrading from `Marker.1` to `Marker.2`: https://github.com/PixarAnimationStudios/OpenTimelineIO/blob/cc78a7c5e9b10808d524edd00a09f8a11b21ce1a/opentimelineio/schema/marker.py#L124-L128

In the end of the day, it may not be best for us to dictate when implementations chose to upconvert to a new schema version or if they should or should not down convert to an older one when there would be no loss in data to do so (e.x. none of the attributes added in a new version were used).

I think out of scope of the specification discussion but still relevant to ensure our specification is pragmatic, the current behavior of failing import on an unsupported schema object may not be the most desirable. The API consumer should likely have the ability to decide how they want to address those issues, either through deserializing as unsupported schema objects and/or through an import log: https://github.com/PixarAnimationStudios/OpenTimelineIO/issues/28

For the moment, too new of schema versions floating around the pipeline has not been much of an issue, though this is likely to change with more DCC adoption of the library/standard. If it does start becoming an issue, it may be fairly well addressed by providing some utilities to assist in downconversion that give users some control over how the potential loss of information is managed in that case.

To recap, my current feeling is that we should:
1. specify the concept of schema up conversion.
2. specify that new schema must always describe incremental up conversion rules from previous schema.
3. NOT specify when implementers should choose to upconvert schema or write older or newer schema, with the exception of specific schema that are marked deprecated (i.e. old schema versions are not implicitly deprecated).
4. When a schema is marked deprecated, it should no longer be written to files.

@ssteinbach and @jminor, does this sound like a good stance to take?

**Comment by palemieux** [comment]
@reinecke Please review #15
